subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/SubscribableBuildActionRunner.java

```diff
@@ -14,83 +14,61 @@
-import org.gradle.tooling.internal.provider.SubscribableBuildAction;
 
 import java.util.ArrayList;
 import java.util.List;
 
 public class SubscribableBuildActionRunner implements BuildActionRunner {
-    private static final BuildOperationListener NO_OP = new BuildOperationListener() {
-        @Override
-        public void started(BuildOperationInternal buildOperation, OperationStartEvent startEvent) {
-        }
-
-        @Override
-        public void finished(BuildOperationInternal buildOperation, OperationResult finishEvent) {
-        }
-    };
     private final BuildActionRunner delegate;
     private final BuildOperationService buildOperationService;
     private final List<BuildOperationListener> listeners = new ArrayList<BuildOperationListener>();
+    private final List<? extends SubscribableBuildActionRunnerRegistration> registrations;
 
-    public SubscribableBuildActionRunner(BuildActionRunner delegate, BuildOperationService buildOperationService) {
+    public SubscribableBuildActionRunner(BuildActionRunner delegate, BuildOperationService buildOperationService, List<? extends SubscribableBuildActionRunnerRegistration> registrations) {
         this.delegate = delegate;
         this.buildOperationService = buildOperationService;
-    }
-
-    private void registerListenersForClientSubscriptions(BuildClientSubscriptions clientSubscriptions, GradleInternal gradle) {
-        BuildEventConsumer eventConsumer = gradle.getServices().get(BuildEventConsumer.class);
-        if (clientSubscriptions.isSendTestProgressEvents()) {
-            registerListener(new ClientForwardingTestOperationListener(eventConsumer, clientSubscriptions));
-        }
-        if (!clientSubscriptions.isSendBuildProgressEvents() && !clientSubscriptions.isSendTaskProgressEvents()) {
-            return;
-        }
-
-        BuildOperationListener buildListener = NO_OP;
-        if (clientSubscriptions.isSendBuildProgressEvents()) {
-            buildListener = new ClientForwardingBuildOperationListener(eventConsumer);
-        }
-        buildListener = new ClientForwardingTaskOperationListener(eventConsumer, clientSubscriptions, buildListener);
-        registerListener(buildListener);
-    }
-
-    private void registerListener(BuildOperationListener listener) {
-        listeners.add(listener);
-        buildOperationService.addListener(listener);
+        this.registrations = registrations;
     }
 
     @Override
     public void run(BuildAction action, BuildController buildController) {
-        if (!(action instanceof SubscribableBuildAction)) {
-            return;
+        boolean subscribable = action instanceof SubscribableBuildAction;
+        if (subscribable) {
+            GradleInternal gradle = buildController.getGradle();
+            SubscribableBuildAction subscribableBuildAction = (SubscribableBuildAction) action;
+            // register listeners that dispatch all progress via the registered BuildEventConsumer instance,
```

Comment:
Yep, this comment was there already .. I'll remove it while addressing your other comments.